### PR TITLE
Fix FSDP when not all outputs get gradient in backward

### PR DESCRIPTION
**Pull request by rohan-varma**
In some use cases, FSDP runs into an issue where a training state assert in `_wait_for_post_backward` erroneously fires. Digging into the root cause, this is because `_post_backward_hook` which sets the module's training state to backward_post is never actually called, since no param in that module had gradient computed for it. Similar to DDP, this can happen when not all module outputs are used in loss computation, or module did not participate in forward at all. 

Fix this by tracking a variable `_post_backward_called` to track whether the hook is actually called or not.

### CODE REVIEW

File: test/distributed/fsdp/test_fsdp_misc.py [hunk 1]
```diff
@@ -49,6 +50,24 @@ def world_size(self):
     def process_group(self):
         return dist.distributed_c10d._get_default_group()
 
+    def test_fsdp_module_no_compute_grad(self):
```
Comment (rohan-varma) [inline test/distributed/fsdp/test_fsdp_misc.py:53]: add no_shard test

File: test/distributed/fsdp/test_fsdp_misc.py [hunk 2]
```diff
@@ -49,6 +50,38 @@ def world_size(self):
...
+        y = torch.randn(10, 10, device='cuda')
+        if use_second_layer:
+            a, b = fsdp(x, y)
+        else:
+            a = fsdp(x, y)
+        loss = a.sum()
+        loss.backward()
+
+        # self.a receives grad, self.b does not
+        a_grad = fsdp._fsdp_wrapped_module._fpw_module.a._fsdp_wrapped_module.flat_param.grad
```
Comment (awgu) [inline test/distributed/fsdp/test_fsdp_misc.py:80]:
```suggestion
        a_grad = fsdp.module.a._fsdp_wrapped_module.flat_param.grad
```

File: test/distributed/fsdp/test_fsdp_misc.py [hunk 3]
```diff
@@ -49,6 +50,38 @@ def world_size(self):
...
+        b_grad = fsdp._fsdp_wrapped_module._fpw_module.b._fsdp_wrapped_module.flat_param.grad
```
Comment (awgu) [inline test/distributed/fsdp/test_fsdp_misc.py:81]:
```suggestion
        b_grad = fsdp.module.b._fsdp_wrapped_module.flat_param.grad
```

File: test/distributed/fsdp/test_fsdp_misc.py [hunk 4]
```diff
@@ -49,6 +50,38 @@ def world_size(self):
...
+        self.assertTrue(a_grad is not None)
+        self.assertTrue(b_grad is None)
```
Comment (awgu) [inline test/distributed/fsdp/test_fsdp_misc.py:83]:
```suggestion
        self.assertIsNone(b_grad)
```

File: test/distributed/fsdp/test_fsdp_misc.py [hunk 5]
```diff
@@ -49,6 +50,38 @@ def world_size(self):
...
+            a, b = fsdp(x, y)
+        else:
+            a = fsdp(x, y)
+        loss = a.sum()
+        loss.backward()
+
+        # self.a receives grad, self.b does not
+        a_grad = fsdp._fsdp_wrapped_module._fpw_module.a._fsdp_wrapped_module.flat_param.grad
+        b_grad = fsdp._fsdp_wrapped_module._fpw_module.b._fsdp_wrapped_module.flat_param.grad
+        self.assertTrue(a_grad is not None)
```
Comment (awgu) [inline test/distributed/fsdp/test_fsdp_misc.py:82]:
```suggestion
        self.assertIsNotNone(a_grad)
```

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 6]
```diff
@@ -3089,24 +3090,39 @@ def _finalize_params(fsdp_module: FullyShardedDataParallel) -> None:
                         p.grad = p._saved_grad_shard  # type: ignore[attr-defined]
                     else:
                         p_assert(
-                            not p._is_sharded, "All sharded parameters should "
+                            not p._is_sharded or not getattr(p, '_post_backward_called', False), "All sharded parameters that received gradient should "
```
Comment (awgu) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:3093]: nit: maybe break this line

Comment (awgu) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:3093]: Should we not set `_post_backward_called = False` for each `FlatParameter` and avoid `getattr(...)`? I.e. enforce  that the attribute is always defined by runtime?
Please correct me if I am wrong, but it seems like we are using whether the attribute exists to indicate whether the post backward was called or not. Hence, the value of the attribute, when defined, is not actually meaningful.
(I do not see anywhere where set `.post_backward_called = False`.)

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 7]
```diff
@@ -3089,24 +3090,39 @@ def _finalize_params(fsdp_module: FullyShardedDataParallel) -> None:
...
+                            getattr(p, '_post_backward_called', False) for p in m.params
+                        )
+                        if hook_called:
+                            m._assert_state(TrainingState_.BACKWARD_POST)
+                            # Delete attr so the logic works correctly next iteration
+                            for p in m.params:
+                                if hasattr(p, '_post_backward_called'):
+                                    delattr(p, '_post_backward_called')
+                        else:
+                            # post backward hook was not called, meaning param
```
Comment (awgu) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:3121]: nit: The comment's grammar assumes single parameter per `FullyShardedDataParallel`, while the code still iterates over `.params`. I prefer that the comment assumes plural parameters per `FullyShardeDataParallel` to match the code and because this will be true for non-recursive wrapping.

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 8]
```diff
@@ -3089,24 +3090,39 @@ def _finalize_params(fsdp_module: FullyShardedDataParallel) -> None:
                         p.grad = p._saved_grad_shard  # type: ignore[attr-defined]
                     else:
                         p_assert(
-                            not p._is_sharded, "All sharded parameters should "
+                            not p._is_sharded or not getattr(p, '_post_backward_called', False), "All sharded parameters that received gradient should "
                             "use `_saved_grad_shard`"
                         )
                     if hasattr(p, "_saved_grad_shard"):
                         delattr(p, "_saved_grad_shard")
+                    if hasattr(p, '_post_backward_called'):
```
Comment (awgu) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:3098]: This seems like a no-op because `_post_backward_called` will have already been deleted before `_finalize_params()` is called.

File: test/distributed/fsdp/test_fsdp_misc.py [hunk 9]
```diff
@@ -49,6 +50,38 @@ def world_size(self):
...
 
+    @parametrize("use_second_layer", [True, False])
+    def test_fsdp_module_no_compute_grad(self, use_second_layer):
+        class MyModel(nn.Module):
+            def __init__(self):
+                super().__init__()
+                self.a = nn.Linear(10, 10)
+                self.b = nn.Linear(10, 10)
+
+            def forward(self, x, y):
```
Comment (awgu) [inline test/distributed/fsdp/test_fsdp_misc.py:61]: `y` is unused? Is this intentional?

File: test/distributed/fsdp/test_fsdp_misc.py [hunk 10]
```diff
@@ -49,6 +50,38 @@ def world_size(self):
     def process_group(self):
         return dist.distributed_c10d._get_default_group()
 
+    @parametrize("use_second_layer", [True, False])
+    def test_fsdp_module_no_compute_grad(self, use_second_layer):
```
Comment (awgu) [inline test/distributed/fsdp/test_fsdp_misc.py:54]: Can we add a comment about the point of `use_second_layer`?

In my understanding:
- If `use_second_layer == False`, then `b` is not involved in the forward computation.
- If `use_second_layer == True`, then `b` is involved in the forward computation but not connected to the loss that initiates the backward.

Before this PR, would both cases have failed?

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 11]
```diff
@@ -3089,24 +3090,39 @@ def _finalize_params(fsdp_module: FullyShardedDataParallel) -> None:
...
-                    if any([p.requires_grad for p in m.params]):
-                        m._assert_state(TrainingState_.BACKWARD_POST)
+                    managed_param_requires_grad = any(p.requires_grad for p in m.params)
+                    if managed_param_requires_grad:
+                        hook_called = any(
+                            getattr(p, '_post_backward_called', False) for p in m.params
+                        )
+                        if hook_called:
+                            m._assert_state(TrainingState_.BACKWARD_POST)
+                            # Delete attr so the logic works correctly next iteration
```
Comment (awgu) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:3116]: ditto to above: maybe this can be changed to having `_post_backward_called` be either `False` or `True`, and then here it would be reset to `False`.

**Review by awgu — COMMENTED** [review]
Leaving some initial comments. Feel free to re-request review.

File: test/distributed/fsdp/test_fsdp_misc.py (hunk 9 above)
Comment (rohan-varma) [reply test/distributed/fsdp/test_fsdp_misc.py:61]: Good catch, should be self.b(y)

File: test/distributed/fsdp/test_fsdp_misc.py (hunk 10 above)
Comment (rohan-varma) [reply test/distributed/fsdp/test_fsdp_misc.py:54]: Will clarify, but yes. Both cases would've failed

File: torch/distributed/fsdp/fully_sharded_data_parallel.py (hunk 6 above)
Comment (rohan-varma) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:3093]: Yeah, we should just always set it to be cleaner.

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 12]
```diff
@@ -1572,6 +1572,14 @@ def _lazy_init(self) -> None:
             # ``optim.step()`` is done before we all-gather parameters.
             self._wait_for_previous_optim_step()
 
+            # Set a flag on every FlatParameter to track whether its post
+            # backward hook has been called, mainly for validation purpose
+            # in wait_for_post_backward.
+            if self._exec_order_data.is_first_iter:
```
Comment (awgu) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:1578]: Note to myself: I will think about where to move this logic with the future refactors.

**Review by awgu — APPROVED** [review]
LGTM!

File: torch/distributed/fsdp/fully_sharded_data_parallel.py (hunk 12 above)
Comment (rohan-varma) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:1578]: Thank you! We can probably adjust this further when lazy init cleanup lands: https://github.com/pytorch/pytorch/pull/80185

**Comment by rohan-varma** [comment]
In progress job actually seems finished: https://github.com/pytorch/pytorch/runs/7081928791?check_suite_focus=true

**Comment by awgu** [comment]
For bookkeeping: I think that https://github.com/pytorch/pytorch/pull/83195 covers the issue tested in this PR's `test_fsdp_module_no_compute_grad(use_second_layer=True)` but fixes the functionality more thoroughly.

This PR's unit test only checked that the grad was `None` or not but not other correctness like parameters being resharded. We can prefer the newer PR's unit test.
